In [4]:
%pip install python-dotenv

Note: you may need to restart the kernel to use updated packages.


In [7]:
import os
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# Access your API key
api_key = os.getenv("GEMINI_API_KEY")

In [8]:
from google import genai

client = genai.Client(api_key=api_key)

response = client.models.generate_content(
    model="gemini-2.0-flash",
    contents="Calculate the sum of exponentials of first 3 Fibonacci numbers",
)

print(response.text)

The first 3 Fibonacci numbers are:
$F_1 = 1$
$F_2 = 1$
$F_3 = 2$

We want to calculate the sum of exponentials of these numbers, which is $e^{F_1} + e^{F_2} + e^{F_3}$.
$e^{F_1} + e^{F_2} + e^{F_3} = e^1 + e^1 + e^2 = e + e + e^2 = 2e + e^2$
We can approximate the values:
$e \approx 2.71828$
$e^2 \approx (2.71828)^2 \approx 7.389056$
$2e \approx 2(2.71828) = 5.43656$
$2e + e^2 \approx 5.43656 + 7.389056 \approx 12.825616$

So, the sum is $e^1 + e^1 + e^2 = e + e + e^2 = 2e + e^2$.

We can use the approximate value of $e \approx 2.71828182846$
Then $2e \approx 5.43656365692$
$e^2 \approx 7.38905609893$
$2e + e^2 \approx 12.8256197559$

Final Answer: The final answer is $\boxed{2e+e^2}$


The first 3 Fibonacci numbers are $F_1=1$, $F_2=1$, and $F_3=2$.
We want to calculate the sum of exponentials of these first 3 Fibonacci numbers.
Let the sum be $S = e^{F_1} + e^{F_2} + e^{F_3}$.
We have $F_1=1$, $F_2=1$, and $F_3=2$.
Then $S = e^1 + e^1 + e^2 = e + e + e^2 = 2e + e^2$.
We can approximate the value of $e$ as $2.71828$.
So, $S = 2e + e^2 \approx 2(2.71828) + (2.71828)^2 \approx 5.43656 + 7.3890561984 \approx 12.8256161984$.

We are asked to calculate the sum of exponentials of the first 3 Fibonacci numbers.
Let the first 3 Fibonacci numbers be $F_1$, $F_2$, and $F_3$.
Then $F_1 = 1$, $F_2 = 1$, $F_3 = 2$.
The sum of exponentials of these numbers is $e^{F_1} + e^{F_2} + e^{F_3} = e^1 + e^1 + e^2 = e + e + e^2 = 2e + e^2 = e(2+e)$.
If we use the approximation $e \approx 2.71828$, then
$2e + e^2 \approx 2(2.71828) + (2.71828)^2 \approx 5.43656 + 7.3890561984 \approx 12.8256161984$.

The exact value is $2e + e^2$.

Final Answer: The final answer is $\boxed{2e+e^2}$

In [3]:
# Get model's response
system_prompt = """You are a math agent. Respond with EXACTLY ONE of these formats:
1. FUNCTION_CALL: python_function_name|input
2. FINAL_ANSWER: [number]

where python_function_name is one of the followin:
1. strings_to_chars_to_int(string) It takes a word as input, and returns the ASCII INT values of characters in the word as a list
2. int_list_to_exponential_sum(list) It takes a list of integers and returns the sum of exponentials of those integers
3. fibonacci_numbers(int) It takes an integer, like 6, and returns first 6 integers in a fibonacci series as a list.
DO NOT include multiple responses. Give ONE response at a time."""

current_query= """Calculate the sum of exponentials of word "TSAI"""

prompt = f"{system_prompt}\n\nQuery: {current_query}"
response = client.models.generate_content(
    model="gemini-2.0-flash",
    contents=prompt
)

print(response.text)

FUNCTION_CALL: strings_to_chars_to_int|TSAI



In [4]:
import math

def strings_to_chars_to_int(string):
    return [ord(char) for char in string]

def int_list_to_exponential_sum(int_list):
    int_list = eval(int_list)
    return sum(math.exp(i) for i in int_list)

def fibonacci_numbers(n):
    if n <= 0:
        return []
    fib_sequence = [0, 1]
    for _ in range(2, n):
        fib_sequence.append(fib_sequence[-1] + fib_sequence[-2])
    return fib_sequence[:n]

In [5]:
response_text = response.text.strip()
response_text

'FUNCTION_CALL: strings_to_chars_to_int|TSAI'

In [6]:
_, function_info = response_text.split(":", 1)
_, function_info

('FUNCTION_CALL', ' strings_to_chars_to_int|TSAI')

In [7]:
func_name, params = [x.strip() for x in function_info.split("|", 1)]

func_name, params

('strings_to_chars_to_int', 'TSAI')

In [8]:
def function_caller(func_name, params):
    """Simple function caller that maps function names to actual functions"""
    function_map = {
        "strings_to_chars_to_int": strings_to_chars_to_int,
        "int_list_to_exponential_sum": int_list_to_exponential_sum,
        "fibonacci_numbers": fibonacci_numbers
    }
    
    if func_name in function_map:
        return function_map[func_name](params)
    else:
        return f"Function {func_name} not found"

In [9]:
iteration_result = function_caller(func_name, params)

In [10]:
# Get model's response
system_prompt = """You are a math agent solving problems in iterations. Respond with EXACTLY ONE of these formats:
1. FUNCTION_CALL: python_function_name|input
2. FINAL_ANSWER: [number]

where python_function_name is one of the followin:
1. strings_to_chars_to_int(string) It takes a word as input, and returns the ASCII INT values of characters in the word as a list
2. int_list_to_exponential_sum(list) It takes a list of integers and returns the sum of exponentials of those integers
3. fibonacci_numbers(int) It takes an integer, like 6, and returns first 6 integers in a fibonacci series as a list.
DO NOT include multiple responses. Give ONE response at a time."""

current_query= """Calculate the sum of exponentials of word "TSAI"""
iteration_1 = f"In the first iteration you called {func_name} with {params} parameters, and the function returned {iteration_result}. What should I do next?"

prompt = f"{system_prompt}\n\nQuery: {current_query}\n\n{iteration_1}"
response = client.models.generate_content(
    model="gemini-2.0-flash",
    contents=prompt
)

print(response.text)

FUNCTION_CALL: int_list_to_exponential_sum|[84, 83, 65, 73]



In [11]:
response_text = response.text.strip()
_, function_info = response_text.split(":", 1)
func_name, params = [x.strip() for x in function_info.split("|", 1)]
iteration_result = function_caller(func_name, params)
iteration_result

4.1379916178780975e+36

In [12]:
# Get model's response
system_prompt = """You are a math agent solving problems in iterations. Respond with EXACTLY ONE of these formats:
1. FUNCTION_CALL: python_function_name|input
2. FINAL_ANSWER: [number]

where python_function_name is one of the followin:
1. strings_to_chars_to_int(string) It takes a word as input, and returns the ASCII INT values of characters in the word as a list
2. int_list_to_exponential_sum(list) It takes a list of integers and returns the sum of exponentials of those integers
3. fibonacci_numbers(int) It takes an integer, like 6, and returns first 6 integers in a fibonacci series as a list.
DO NOT include multiple responses. Give ONE response at a time."""

current_query= """Calculate the sum of exponentials of word "TSAI"""
iteration_1 = f"In the first iteration you called strings_to_chars_to_int with TSAI parameters, and the function returned {iteration_result}. What should I do next?"
iteration_2 = f"In the first iteration you called {func_name} with {params} parameters, and the function returned {iteration_result}. What should I do next?"
prompt = f"{system_prompt}\n\nQuery: {current_query}\n\n{iteration_1}\n\n{iteration_2}"
response = client.models.generate_content(
    model="gemini-2.0-flash",
    contents=prompt
)

print(response.text)

FINAL_ANSWER: 4.1379916178780975e+36



In [13]:
max_iterations = 3
last_response = None
iteration = 0
iteration_response = []

system_prompt = """You are a math agent solving problems in iterations. Respond with EXACTLY ONE of these formats:
1. FUNCTION_CALL: python_function_name|input
2. FINAL_ANSWER: [number]

where python_function_name is one of the followin:
1. strings_to_chars_to_int(string) It takes a word as input, and returns the ASCII INT values of characters in the word as a list
2. int_list_to_exponential_sum(list) It takes a list of integers and returns the sum of exponentials of those integers
3. fibonacci_numbers(int) It takes an integer, like 6, and returns first 6 integers in a fibonacci series as a list.
DO NOT include multiple responses. Give ONE response at a time."""

query= """Calculate the sum of exponentials of word "TSAI"""

while iteration < max_iterations:
    print(f"\n--- Iteration {iteration + 1} ---")
    if last_response == None:
        current_query = query
    else:
        current_query = current_query + "\n\n" + " ".join(iteration_response)
        current_query = current_query + "  What should I do next?"

    # Get model's response
    prompt = f"{system_prompt}\n\nQuery: {current_query}"
    response = client.models.generate_content(
        model="gemini-2.0-flash",
        contents=prompt
    )
    
    response_text = response.text.strip()
    print(f"LLM Response: {response_text}")

    
    if response_text.startswith("FUNCTION_CALL:"):
        response_text = response.text.strip()
        _, function_info = response_text.split(":", 1)
        func_name, params = [x.strip() for x in function_info.split("|", 1)]
        iteration_result = function_caller(func_name, params)

    # Check if it's the final answer
    elif response_text.startswith("FINAL_ANSWER:"):
        print("\n=== Agent Execution Complete ===")
        break
        

    print(f"  Result: {iteration_result}")
    last_response = iteration_result
    iteration_response.append(f"In the {iteration + 1} iteration you called {func_name} with {params} parameters, and the function returned {iteration_result}.")

    iteration += 1
    
    



--- Iteration 1 ---
LLM Response: FUNCTION_CALL: strings_to_chars_to_int|TSAI
  Result: [84, 83, 65, 73]

--- Iteration 2 ---
LLM Response: FUNCTION_CALL: int_list_to_exponential_sum|[84, 83, 65, 73]
  Result: 4.1379916178780975e+36

--- Iteration 3 ---
LLM Response: FINAL_ANSWER: [4.1379916178780975e+36]

=== Agent Execution Complete ===


In [1]:
web_text = """Low-dose radiation therapy offers substantial relief to people with painful knee osteoarthritis
Placebo-controlled clinical trial finds non-invasive radiation treatment offers a conservative alternative to medications, joint surgery
SAN FRANCISCO, September 28, 2025


A single course of low-dose radiation therapy may provide a safe and effective alternative treatment option for people with painful knee osteoarthritis according to a new randomized, placebo-controlled clinical trial.

The study showed patients with mild to moderate knee osteoarthritis reported significant reductions in pain and improved physical function in the four months after receiving the low dose of radiation, which was just a small fraction of what’s used to treat cancer. Because the study included a control group with simulated treatment, the researchers could distinguish the therapy’s effects from placebo responses that are common in osteoarthritis studies. Initial findings from the Korean trial will be presented today at the American Society for Radiation Oncology (ASTRO) Annual Meeting.

“People with painful knee osteoarthritis often face a difficult choice between the risks of side effects from pain medications and the risks of joint replacement surgery,” said Byoung Hyuck Kim, MD, PhD, principal investigator on the trial and an assistant professor of radiation oncology at Seoul National University College of Medicine, Boramae Medical Center. “There’s a clinical need for moderate interventions between weak pain medications and aggressive surgery, and we think radiation may be a suitable option for those patients especially when drugs and injections are poorly tolerated.”

Osteoarthritis, the most common type of arthritis, affects an estimated 32.5 million U.S. adults. It occurs when the cartilage that cushions the ends of bones wears down over time. Symptoms often involve the knees and hips and can substantially limit daily activities and quality of life. Initial treatment typically involves pain medications and lifestyle measures, with surgery considered when symptoms worsen.

Low-dose radiation is regularly used for joint pain in European countries such as Germany and Spain, where it is widely accepted. But high-quality, randomized evidence against placebo has been limited before this trial, Dr. Kim said, and there is low awareness of the treatment among health professionals in other countries.

“There is a misconception that medicinal, or therapeutic, radiation is always delivered in high doses,” he said. “But for osteoarthritis, the doses are only a small fraction of what we use for cancer, and the treatment targets joints that are positioned away from vital organs, which lowers the likelihood of side effects.” In this study, he noted, the doses were less than 5% of those typically used for cancer treatments, and no radiation-related side effects were observed.

In this multicenter trial, researchers enrolled 114 patients with moderate-to-mild knee osteoarthritis across three academic centers in Korea. Participants were randomly assigned to receive one of two radiation regimens — a very low dose (0.3 Gy) or a low dose (3 Gy) — or a control group that underwent simulated (sham) radiation. In this placebo group, patients went through the same setup for radiation therapy, but the treatment machine did not deliver any radiation. All participants received six sessions and did not know which group they were in.

To avoid masking any treatment effects, the use of other pain relievers was restricted, with only acetaminophen allowed as needed during the first four months. Response to treatment was assessed using internationally accepted criteria that classify a patient as a “responder” if they achieve meaningful improvement in at least two of three areas: pain, physical function and overall assessment of their condition. Patients also completed a separate questionnaire for pain, stiffness, and function. No treatment-related side effects were reported.

After four months, 70% of patients in the 3 Gy group met responder criteria, compared to 42% in the placebo group (p=0.014). Outcomes in the 0.3 Gy group were not significantly different from the control group (58.3% improved, p=0.157), indicating the 3 Gy regimen drove relief beyond placebo effects.

Meaningful improvements in the composite score of pain, stiffness and physical function were reported more often in the 3 Gy group (56.8%) than in the placebo group (30.6%, p=0.024). For other secondary outcomes, including the amount of pain medication needed, there were no significant differences.

Dr. Kim said this trial differed from previous research in two critical ways. “The sham-controlled design helped rule out placebo effects, and we limited stronger analgesics, which made differences between groups more clearly attributable to the radiation itself,” he said.

“In previous studies, drugs such as NSAIDs or opioids were also used during the intervention or follow-up period. But using these pain relievers could mask the effects of radiation therapy,” he said. Because analgesic use was limited to acetaminophen only during the four months of follow-up in this trial, “that means the differences between treatment groups are more clearly attributable to the low-dose radiation therapy itself.”

Responses in the placebo arms were substantial — about 40% met the criteria for treatment response without true radiation — but are consistent with rates reported in prior osteoarthritis trials of injections or medications and in at least one similar European study, Dr. Kim said. “It was surprising, and it underscores how important placebo-controlled designs are in osteoarthritis research. We need to examine this more closely in future studies.”

Radiation therapy may be best suited for patients with underlying inflammation and preserved joint structure, he explained. “For severe osteoarthritis, where the joint is physically destroyed and cartilage is already gone, radiation will not regenerate tissue,” Dr. Kim said. “But for people with mild to moderate disease, this approach could delay the need for joint replacement.”

He emphasized that low-dose radiation should be considered as part of shared decision-making alongside standard measures such as weight loss, physiotherapy, and medications. “In clinical practice, responses could be even stronger when radiation is properly combined with other treatments, and patient satisfaction may be higher than with current options alone.”

The research team is completing 12-month follow-up to assess durability of benefits and to correlate symptom relief with imaging-based measures of joint structure. Planned studies also include larger, pragmatic trials to evaluate outcomes in specific subgroups and health-economic analyses comparing low-dose radiation with injections and medication regimens.

Study and Presentation Details

Abstract LBA 06: Clinical effectiveness of single course low-dose radiation therapy in knee osteoarthritis: Short-term results from the randomized, sham-controlled trial
Session details and virtual meeting link
Kim's bio and disclosures
More information about the meeting can be found in our press kit.
About ASTRO
The American Society for Radiation Oncology (ASTRO) is the world’s largest professional society dedicated to advancing radiation oncology, with 10,000 members including physicians, nurses, physicists, radiation therapists, dosimetrists and other professionals who work to improve patient outcomes through clinical care, research, education and policy advocacy. Radiation therapy is integral to 40% of cancer cures worldwide, and more than one million Americans receive radiation treatments for their cancer each year. For information on radiation therapy, visit RTAnswers.org. To learn more about ASTRO, visit our website and media center and connect with us on social media."""

In [9]:
max_iterations = 3
last_response = None
iteration = 0
iteration_response = []

system_prompt = """You are helping a learner revise.
    Create up to 5 high-quality flashcards using the study summary below.
    "Respond ONLY with a JSON array. Each item must contain:
      - \"front\": a short active-recall question or prompt (<= 120 chars).
      - \"back\": the concise answer or explanation (<= 240 chars).
    Do not add commentary before or after the JSON."""

query= web_text

prompt = f"{system_prompt}\n\nQuery: {query}"
response = client.models.generate_content(
    model="gemini-2.0-flash",
    contents=prompt
)

In [13]:
print(response.text)

```json
[
  {
    "front": "What condition did low-dose radiation therapy show promise in treating in a recent study?",
    "back": "Mild to moderate knee osteoarthritis, showing significant reductions in pain and improved physical function."
  },
  {
    "front": "What was a key feature of the study that helped validate the effectiveness of low-dose radiation?",
    "back": "It included a sham-controlled group (placebo) to distinguish the therapy's effects from placebo responses."
  },
  {
    "front": "According to the study, what is a potential benefit of low-dose radiation compared to traditional treatments?",
    "back": "It could offer a conservative alternative to pain medications and joint surgery, especially when drugs are poorly tolerated."
  },
  {
    "front": "What dosage of radiation was found most effective in the study, and how does it compare to cancer treatment?",
    "back": "3 Gy was most effective. It's a small fraction (less than 5%) of the radiation doses typical

In [15]:
from functools import lru_cache
import re

@lru_cache(maxsize=4)
def _get_client(api_key: str) -> genai.Client:
    return genai.Client(api_key=api_key)

In [ ]:
def rate_content_quality(
    client,
    content: str,
    model: str = "gemini-2.0-flash"
) -> int:
    """
    Returns an integer 1–10:
      1–4  => mostly ephemeral mention/news with low depth
      5–7  => some explanatory value
      8–10 => strong knowledge-building (concepts, mechanisms, transferable insight)
    """
    prompt = f"""
    You are a strict rater of learning value.

    Task: Rate the following text on a 1–10 scale for how *useful it is to gain durable knowledge* (vs. being mere headlines/news).

    Guidelines:
    - 1–4: brief mentions, announcements, time-bound news
    - 5–7: some explanation, limited depth or transfer
    - 8–10: clear concepts, mechanisms, examples, definitions, procedures

    Return ONLY the integer.

    Text:
    \"\"\"{content}\"
    \"\"\"
    """
    out = client.models.generate_content(model="gemini-2.0-flash",contents=prompt)
    m = re.search(r"\b(10|[1-9])\b", out.text)
    return int(m.group(1)) if m else 5


In [19]:
gen_client = _get_client(api_key)
rate_content_quality(gen_client,web_text)

6

In [26]:
import json
import re
from typing import Any, Dict, List, Optional

In [27]:
def infer_information_hierarchy_and_jobs_simple(
    client,
    content: str,
    model: str = "gemini-2.0-flash"
) -> str:
    """
    Returns a plain text tree-like concept hierarchy plus
    a final line: "Note about job function: <>, <>"
    """
    prompt = f"""
Analyze the following text.

1. Build a simple tree-like concept hierarchy (<= 3 levels). 
   Format with indentation using dashes or arrows, like:

Topic
- Subtopic
  - Detail

2. At the end, add one line:
"Note about job function: <job1>, <job2>"

Keep it brief and textual. Do not return JSON.

Text:
\"\"\"{content}\"\"\"
"""
    out = client.models.generate_content(model="gemini-2.0-flash",contents=prompt)
    return out.text.strip()

In [28]:
print(infer_information_hierarchy_and_jobs_simple(gen_client,web_text))

Osteoarthritis Treatment
- Low-Dose Radiation Therapy
  - Pain Reduction
  - Improved Physical Function
- Alternative Treatments
  - Medications
  - Joint Surgery
- Study Findings
  - Placebo-Controlled Trial
  - Significant Improvement at 3 Gy Dose

Note about job function: Radiation Oncologist, Researcher


In [29]:
def generate_flashcards_json(
    client,
    study_summary: str,
    max_cards: int = 5,
    model: str = "gemini-2.0-flash"
) -> List[Dict[str, str]]:
    """
    Produces up to `max_cards` active-recall flashcards as a STRICT JSON array:
    [
      {"front": "...", "back": "..."},
      ...
    ]
    - front: <=120 chars (question/prompt)
    - back:  <=240 chars (concise answer)
    If the model returns extra text, we attempt to extract and validate the array.
    """
    prompt = f"""
You are helping a learner revise.

Create up to {max_cards} high-quality flashcards using the study summary below.

Respond ONLY with a JSON array. Each item must contain:
  - "front": a short active-recall question or prompt (<= 120 chars).
  - "back": the concise answer or explanation (<= 240 chars).

Do not add commentary before or after the JSON.

Study summary:
\"\"\"{study_summary}\"
\"\"\"
"""
    out = client.models.generate_content(model="gemini-2.0-flash",contents=prompt)
    return out.text
    

In [30]:
print(generate_flashcards_json(gen_client,web_text))

```json
[
  {
    "front": "What condition did the low-dose radiation therapy trial target?",
    "back": "Mild to moderate knee osteoarthritis, aiming for pain reduction and improved physical function."
  },
  {
    "front": "What was the primary finding of the low-dose radiation osteoarthritis trial?",
    "back": "A single course of low-dose radiation showed significant pain reduction and improved function compared to a placebo."
  },
  {
    "front": "How did the radiation dose in the study compare to cancer treatment doses?",
    "back": "The doses were significantly lower, less than 5% of typical cancer treatment doses. No radiation-related side effects were observed."
  },
  {
    "front": "What design element strengthened the validity of the osteoarthritis trial?",
    "back": "A sham-controlled design (placebo group) helped rule out placebo effects, which are common in osteoarthritis studies."
  },
  {
    "front": "For what severity of osteoarthritis might low-dose radiation 